File: src/orchestrator/core/validator-registry.ts
import { spawn } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import type {
  AcceptanceCriterion,
  AcceptanceCriterionExecutionReport,
  VerificationSpecType,
} from '../mission/types';
import type { CommandResult, VerificationRunner } from '../verification-runner';
import type { DispatchBatch } from './dispatch/dispatch-batch';

export interface VerificationCommandOptions {
  cwd?: string;
  timeoutMs?: number;
}

export interface VerificationCommandRunner {
  (command: string, options?: VerificationCommandOptions): Promise<CommandResult>;
}

export interface VerificationCustomValidatorResult {
  passed: boolean;
  detail: string;
}

export interface VerificationCustomValidator {
  (
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): Promise<VerificationCustomValidatorResult> | VerificationCustomValidatorResult;
}

export interface VerificationSpecExecutionContext {
  workspaceRoot: string;
  batch?: DispatchBatch;
  modifiedFiles?: string[];
  verificationRunner?: VerificationRunner;
  runCommand?: VerificationCommandRunner;
  customValidators?: Record<string, VerificationCustomValidator>;
}

export type VerificationSpecExecutionResult = AcceptanceCriterionExecutionReport;

export interface VerificationSpecExecutor {
  readonly type: VerificationSpecType;
  readonly executorId: string;
  execute(
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): Promise<VerificationSpecExecutionResult> | VerificationSpecExecutionResult;
}

function buildFailureResult(
  criterion: AcceptanceCriterion,
  type: VerificationSpecType,
  detail: string,
  executorId: string,
): VerificationSpecExecutionResult {
  return {
    criterionId: criterion.id,
    type,
    status: 'failed',
    detail,
    executorId,
  };
}

class FileExistsVerificationExecutor implements VerificationSpecExecutor {
  readonly type = 'file_exists' as const;
  readonly executorId = 'builtin:file_exists';

  execute(
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): VerificationSpecExecutionResult {
    const spec = criterion.verificationSpec;
    if (!spec?.targetPath) {
      return buildFailureResult(criterion, this.type, 'file_exists: targetPath 未指定', this.executorId);
    }

    const fullPath = path.resolve(context.workspaceRoot, spec.targetPath);
    const exists = fs.existsSync(fullPath);
    return {
      criterionId: criterion.id,
      type: this.type,
      status: exists ? 'passed' : 'failed',
      detail: exists ? `文件存在: ${spec.targetPath}` : `文件不存在: ${spec.targetPath}`,
      executorId: this.executorId,
    };
  }
}

class FileContentVerificationExecutor implements VerificationSpecExecutor {
  readonly type = 'file_content' as const;
  readonly executorId = 'builtin:file_content';

  execute(
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): VerificationSpecExecutionResult {
    const spec = criterion.verificationSpec;
    if (!spec?.targetPath || !spec.expectedContent) {
      return buildFailureResult(
        criterion,
        this.type,
        'file_content: targetPath 或 expectedContent 未指定',
        this.executorId,
      );
    }

    const filePath = path.resolve(context.workspaceRoot, spec.targetPath);
    if (!fs.existsSync(filePath)) {
      return buildFailureResult(criterion, this.type, `文件不存在: ${spec.targetPath}`, this.executorId);
    }

    const content = fs.readFileSync(filePath, 'utf-8');
    const mode = spec.contentMatchMode || 'contains';
    let matched = false;

    if (mode === 'exact') {
      matched = content === spec.expectedContent;
    } else if (mode === 'contains') {
      matched = content.includes(spec.expectedContent);
    } else {
      matched = new RegExp(spec.expectedContent).test(content);
    }

    return {
      criterionId: criterion.id,
      type: this.type,
      status: matched ? 'passed' : 'failed',
      detail: matched
        ? `文件内容匹配(${mode}): ${spec.targetPath}`
        : `文件内容不匹配(${mode}): ${spec.targetPath}`,
      executorId: this.executorId,
    };
  }
}

class TestPassVerificationExecutor implements VerificationSpecExecutor {
  readonly type = 'test_pass' as const;
  readonly executorId = 'builtin:test_pass';

  async execute(
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): Promise<VerificationSpecExecutionResult> {
    const spec = criterion.verificationSpec;
    const command = spec?.testCommand?.trim();
    if (!command) {
      return buildFailureResult(criterion, this.type, 'test_pass: testCommand 未指定', this.executorId);
    }
    if (!context.runCommand) {
      return buildFailureResult(criterion, this.type, 'test_pass: runCommand 未提供', this.executorId);
    }

    const result = await context.runCommand(command, { cwd: context.workspaceRoot });
    const failureOutput = compactCommandOutput(result.error || result.output);
    return {
      criterionId: criterion.id,
      type: this.type,
      status: result.success ? 'passed' : 'failed',
      detail: result.success
        ? `测试通过: ${command}`
        : `测试失败: ${command}${failureOutput ? ` | ${failureOutput}` : ''}`,
      executorId: this.executorId,
    };
  }
}

class TaskCompletedVerificationExecutor implements VerificationSpecExecutor {
  readonly type = 'task_completed' as const;
  readonly executorId = 'builtin:task_completed';

  execute(
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): VerificationSpecExecutionResult {
    const pattern = criterion.verificationSpec?.taskPattern?.trim().toLowerCase();
    if (!pattern) {
      return buildFailureResult(criterion, this.type, 'task_completed: taskPattern 未指定', this.executorId);
    }
    if (!context.batch) {
      return buildFailureResult(criterion, this.type, 'task_completed: batch 上下文缺失', this.executorId);
    }

    const matches = context.batch.getEntries().filter((entry) => {
      const texts = [
        entry.taskId,
        entry.taskContract.taskTitle,
        entry.taskContract.requirementAnalysis.goal,
      ]
        .filter((value): value is string => typeof value === 'string' && value.trim().length > 0)
        .map((value) => value.toLowerCase());
      return texts.some((text) => text.includes(pattern));
    });

    if (matches.length === 0) {
      return buildFailureResult(criterion, this.type, `未找到匹配任务: ${pattern}`, this.executorId);
    }

    const incomplete = matches.filter((entry) => entry.status !== 'completed' || entry.result?.success === false);
    if (incomplete.length > 0) {
      return buildFailureResult(
        criterion,
        this.type,
        `匹配任务未全部完成: ${incomplete.map((entry) => `${entry.taskId}=${entry.status}`).join(', ')}`,
        this.executorId,
      );
    }

    return {
      criterionId: criterion.id,
      type: this.type,
      status: 'passed',
      detail: `匹配任务已完成: ${matches.map((entry) => entry.taskId).join(', ')}`,
      executorId: this.executorId,
    };
  }
}

class CustomVerificationExecutor implements VerificationSpecExecutor {
  readonly type = 'custom' as const;
  readonly executorId = 'builtin:custom';

  async execute(
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): Promise<VerificationSpecExecutionResult> {
    const validatorName = criterion.verificationSpec?.customValidator?.trim();
    if (!validatorName) {
      return buildFailureResult(criterion, this.type, 'custom: customValidator 未指定', this.executorId);
    }

    const validator = context.customValidators?.[validatorName];
    if (!validator) {
      return buildFailureResult(criterion, this.type, `custom: 未注册校验器 ${validatorName}`, this.executorId);
    }

    const result = await validator(criterion, context);
    return {
      criterionId: criterion.id,
      type: this.type,
      status: result.passed ? 'passed' : 'failed',
      detail: result.detail,
      executorId: `custom:${validatorName}`,
    };
  }
}

function compactCommandOutput(text: string | undefined, limit: number = 240): string {
  const normalized = typeof text === 'string' ? text.trim().replace(/\s+/g, ' ') : '';
  if (!normalized) {
    return '';
  }
  return normalized.length > limit ? `${normalized.slice(0, limit)}...` : normalized;
}

export function createProcessVerificationCommandRunner(
  defaultOptions?: VerificationCommandOptions,
): VerificationCommandRunner {
  return (command: string, options?: VerificationCommandOptions) => new Promise((resolve) => {
    const startTime = Date.now();
    const cwd = options?.cwd || defaultOptions?.cwd;
    const timeout = options?.timeoutMs || defaultOptions?.timeoutMs || 60_000;
    const process = spawn(command, {
      cwd,
      shell: true,
      timeout,
    });

    let stdout = '';
    let stderr = '';

    process.stdout?.on('data', (data) => {
      stdout += data.toString();
    });

    process.stderr?.on('data', (data) => {
      stderr += data.toString();
    });

    process.on('close', (code) => {
      resolve({
        success: code === 0,
        output: stdout,
        error: code === 0 ? undefined : (stderr || stdout || `命令执行失败，退出码: ${code}`),
        duration: Date.now() - startTime,
      });
    });

    process.on('error', (error) => {
      resolve({
        success: false,
        output: stdout,
        error: error.message,
        duration: Date.now() - startTime,
      });
    });
  });
}

export class ValidatorRegistry {
  private readonly executors = new Map<VerificationSpecType, VerificationSpecExecutor>();

  register(executor: VerificationSpecExecutor, options?: { replace?: boolean }): void {
    const existing = this.executors.get(executor.type);
    if (existing && !options?.replace) {
      throw new Error(`验证器已注册: ${executor.type} -> ${existing.executorId}`);
    }
    this.executors.set(executor.type, executor);
  }

  get(type: VerificationSpecType): VerificationSpecExecutor | undefined {
    return this.executors.get(type);
  }

  async executeCriterion(
    criterion: AcceptanceCriterion,
    context: VerificationSpecExecutionContext,
  ): Promise<VerificationSpecExecutionResult | null> {
    if (!criterion.verifiable || !criterion.verificationSpec) {
      return null;
    }

    const executor = this.executors.get(criterion.verificationSpec.type);
    if (!executor) {
      return buildFailureResult(
        criterion,
        criterion.verificationSpec.type,
        `未知验证类型: ${criterion.verificationSpec.type}`,
        'registry:missing-executor',
      );
    }

    try {
      return await executor.execute(criterion, context);
    } catch (error: any) {
      return buildFailureResult(
        criterion,
        criterion.verificationSpec.type,
        `验证异常: ${error?.message || String(error)}`,
        executor.executorId,
      );
    }
  }

  async executeCriteria(
    criteria: AcceptanceCriterion[],
    context: VerificationSpecExecutionContext,
  ): Promise<VerificationSpecExecutionResult[]> {
    const results: VerificationSpecExecutionResult[] = [];
    for (const criterion of criteria) {
      const result = await this.executeCriterion(criterion, context);
      if (result) {
        results.push(result);
      }
    }
    return results;
  }
}

export function createDefaultValidatorRegistry(): ValidatorRegistry {
  const registry = new ValidatorRegistry();
  registry.register(new FileExistsVerificationExecutor());
  registry.register(new FileContentVerificationExecutor());
  registry.register(new TaskCompletedVerificationExecutor());
  registry.register(new TestPassVerificationExecutor());
  registry.register(new CustomVerificationExecutor());
  return registry;
}

File: crates/magi-orchestrator/src/orchestration/validator_registry.rs
use std::collections::HashMap;
use std::path::Path;

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum VerificationSpecType {
    FileExists,
    FileContent,
    TestPass,
    TaskCompleted,
    Custom,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ContentMatchMode {
    Contains,
    Exact,
    Regex,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationSpec {
    #[serde(rename = "type")]
    pub spec_type: VerificationSpecType,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_content: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub content_match_mode: Option<ContentMatchMode>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub test_command: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub task_pattern: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub custom_validator: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcceptanceCriterion {
    pub id: String,
    pub description: String,
    pub verifiable: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub verification_spec: Option<VerificationSpec>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CriterionExecutionReport {
    pub criterion_id: String,
    #[serde(rename = "type")]
    pub spec_type: VerificationSpecType,
    pub status: String,
    pub detail: String,
    pub executor_id: String,
}

pub struct VerificationContext {
    pub workspace_root: String,
    pub modified_files: Vec<String>,
}

fn build_failure(
    criterion: &AcceptanceCriterion,
    spec_type: VerificationSpecType,
    detail: &str,
    executor_id: &str,
) -> CriterionExecutionReport {
    CriterionExecutionReport {
        criterion_id: criterion.id.clone(),
        spec_type,
        status: "failed".to_string(),
        detail: detail.to_string(),
        executor_id: executor_id.to_string(),
    }
}

fn execute_file_exists(
    criterion: &AcceptanceCriterion,
    ctx: &VerificationContext,
) -> CriterionExecutionReport {
    let spec = criterion.verification_spec.as_ref().unwrap();
    let Some(target) = &spec.target_path else {
        return build_failure(
            criterion,
            VerificationSpecType::FileExists,
            "file_exists: targetPath 未指定",
            "builtin:file_exists",
        );
    };
    let full_path = Path::new(&ctx.workspace_root).join(target);
    let exists = full_path.exists();
    CriterionExecutionReport {
        criterion_id: criterion.id.clone(),
        spec_type: VerificationSpecType::FileExists,
        status: if exists { "passed" } else { "failed" }.to_string(),
        detail: if exists {
            format!("文件存在: {target}")
        } else {
            format!("文件不存在: {target}")
        },
        executor_id: "builtin:file_exists".to_string(),
    }
}

fn execute_file_content(
    criterion: &AcceptanceCriterion,
    ctx: &VerificationContext,
) -> CriterionExecutionReport {
    let spec = criterion.verification_spec.as_ref().unwrap();
    let (Some(target), Some(expected)) = (&spec.target_path, &spec.expected_content) else {
        return build_failure(
            criterion,
            VerificationSpecType::FileContent,
            "file_content: targetPath 或 expectedContent 未指定",
            "builtin:file_content",
        );
    };
    let full_path = Path::new(&ctx.workspace_root).join(target);
    let Ok(content) = std::fs::read_to_string(&full_path) else {
        return build_failure(
            criterion,
            VerificationSpecType::FileContent,
            &format!("文件不存在: {target}"),
            "builtin:file_content",
        );
    };
    let mode = spec
        .content_match_mode
        .as_ref()
        .unwrap_or(&ContentMatchMode::Contains);
    let matched = match mode {
        ContentMatchMode::Exact => content == *expected,
        ContentMatchMode::Contains => content.contains(expected.as_str()),
        ContentMatchMode::Regex => regex::Regex::new(expected)
            .map(|re| re.is_match(&content))
            .unwrap_or(false),
    };
    let mode_label = match mode {
        ContentMatchMode::Exact => "exact",
        ContentMatchMode::Contains => "contains",
        ContentMatchMode::Regex => "regex",
    };
    CriterionExecutionReport {
        criterion_id: criterion.id.clone(),
        spec_type: VerificationSpecType::FileContent,
        status: if matched { "passed" } else { "failed" }.to_string(),
        detail: if matched {
            format!("文件内容匹配({mode_label}): {target}")
        } else {
            format!("文件内容不匹配({mode_label}): {target}")
        },
        executor_id: "builtin:file_content".to_string(),
    }
}

pub struct ValidatorRegistry {
    executors: HashMap<VerificationSpecType, String>,
}

impl ValidatorRegistry {
    pub fn new() -> Self {
        let mut executors = HashMap::new();
        executors.insert(
            VerificationSpecType::FileExists,
            "builtin:file_exists".to_string(),
        );
        executors.insert(
            VerificationSpecType::FileContent,
            "builtin:file_content".to_string(),
        );
        executors.insert(
            VerificationSpecType::TaskCompleted,
            "builtin:task_completed".to_string(),
        );
        executors.insert(
            VerificationSpecType::TestPass,
            "builtin:test_pass".to_string(),
        );
        executors.insert(
            VerificationSpecType::Custom,
            "builtin:custom".to_string(),
        );
        Self { executors }
    }

    pub fn register(&mut self, spec_type: VerificationSpecType, executor_id: &str) {
        self.executors.insert(spec_type, executor_id.to_string());
    }

    pub fn has_executor(&self, spec_type: &VerificationSpecType) -> bool {
        self.executors.contains_key(spec_type)
    }

    pub fn execute_criterion(
        &self,
        criterion: &AcceptanceCriterion,
        ctx: &VerificationContext,
    ) -> Option<CriterionExecutionReport> {
        if !criterion.verifiable {
            return None;
        }
        let spec = criterion.verification_spec.as_ref()?;

        if !self.executors.contains_key(&spec.spec_type) {
            return Some(build_failure(
                criterion,
                spec.spec_type,
                &format!("未知验证类型: {:?}", spec.spec_type),
                "registry:missing-executor",
            ));
        }

        Some(match spec.spec_type {
            VerificationSpecType::FileExists => execute_file_exists(criterion, ctx),
            VerificationSpecType::FileContent => execute_file_content(criterion, ctx),
            VerificationSpecType::TestPass => build_failure(
                criterion,
                VerificationSpecType::TestPass,
                "test_pass: 需要外部命令执行器",
                "builtin:test_pass",
            ),
            VerificationSpecType::TaskCompleted => build_failure(
                criterion,
                VerificationSpecType::TaskCompleted,
                "task_completed: 需要 batch 上下文",
                "builtin:task_completed",
            ),
            VerificationSpecType::Custom => build_failure(
                criterion,
                VerificationSpecType::Custom,
                "custom: 需要外部校验器",
                "builtin:custom",
            ),
        })
    }

    pub fn execute_criteria(
        &self,
        criteria: &[AcceptanceCriterion],
        ctx: &VerificationContext,
    ) -> Vec<CriterionExecutionReport> {
        criteria
            .iter()
            .filter_map(|c| self.execute_criterion(c, ctx))
            .collect()
    }
}

impl Default for ValidatorRegistry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn temp_workspace() -> tempfile::TempDir {
        tempfile::tempdir().unwrap()
    }

    fn make_criterion(
        id: &str,
        spec_type: VerificationSpecType,
        target_path: Option<&str>,
        expected_content: Option<&str>,
    ) -> AcceptanceCriterion {
        AcceptanceCriterion {
            id: id.to_string(),
            description: "test".to_string(),
            verifiable: true,
            verification_spec: Some(VerificationSpec {
                spec_type,
                target_path: target_path.map(|s| s.to_string()),
                expected_content: expected_content.map(|s| s.to_string()),
                content_match_mode: None,
                test_command: None,
                task_pattern: None,
                custom_validator: None,
            }),
        }
    }

    #[test]
    fn file_exists_pass() {
        let dir = temp_workspace();
        let file_path = dir.path().join("test.txt");
        std::fs::File::create(&file_path).unwrap();

        let registry = ValidatorRegistry::new();
        let criterion = make_criterion("c1", VerificationSpecType::FileExists, Some("test.txt"), None);
        let ctx = VerificationContext {
            workspace_root: dir.path().to_string_lossy().to_string(),
            modified_files: vec![],
        };
        let result = registry.execute_criterion(&criterion, &ctx).unwrap();
        assert_eq!(result.status, "passed");
    }

    #[test]
    fn file_exists_fail() {
        let dir = temp_workspace();
        let registry = ValidatorRegistry::new();
        let criterion =
            make_criterion("c1", VerificationSpecType::FileExists, Some("missing.txt"), None);
        let ctx = VerificationContext {
            workspace_root: dir.path().to_string_lossy().to_string(),
            modified_files: vec![],
        };
        let result = registry.execute_criterion(&criterion, &ctx).unwrap();
        assert_eq!(result.status, "failed");
    }

    #[test]
    fn file_content_contains_pass() {
        let dir = temp_workspace();
        let file_path = dir.path().join("hello.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        f.write_all(b"hello world foo bar").unwrap();

        let registry = ValidatorRegistry::new();
        let criterion = make_criterion(
            "c2",
            VerificationSpecType::FileContent,
            Some("hello.txt"),
            Some("world foo"),
        );
        let ctx = VerificationContext {
            workspace_root: dir.path().to_string_lossy().to_string(),
            modified_files: vec![],
        };
        let result = registry.execute_criterion(&criterion, &ctx).unwrap();
        assert_eq!(result.status, "passed");
    }

    #[test]
    fn file_content_not_found() {
        let dir = temp_workspace();
        let registry = ValidatorRegistry::new();
        let criterion = make_criterion(
            "c3",
            VerificationSpecType::FileContent,
            Some("nope.txt"),
            Some("content"),
        );
        let ctx = VerificationContext {
            workspace_root: dir.path().to_string_lossy().to_string(),
            modified_files: vec![],
        };
        let result = registry.execute_criterion(&criterion, &ctx).unwrap();
        assert_eq!(result.status, "failed");
        assert!(result.detail.contains("不存在"));
    }

    #[test]
    fn non_verifiable_returns_none() {
        let registry = ValidatorRegistry::new();
        let criterion = AcceptanceCriterion {
            id: "c4".to_string(),
            description: "not verifiable".to_string(),
            verifiable: false,
            verification_spec: None,
        };
        let ctx = VerificationContext {
            workspace_root: "/tmp".to_string(),
            modified_files: vec![],
        };
        assert!(registry.execute_criterion(&criterion, &ctx).is_none());
    }

    #[test]
    fn execute_criteria_batch() {
        let dir = temp_workspace();
        let file_path = dir.path().join("a.txt");
        std::fs::File::create(&file_path).unwrap();

        let registry = ValidatorRegistry::new();
        let criteria = vec![
            make_criterion("c1", VerificationSpecType::FileExists, Some("a.txt"), None),
            make_criterion("c2", VerificationSpecType::FileExists, Some("b.txt"), None),
        ];
        let ctx = VerificationContext {
            workspace_root: dir.path().to_string_lossy().to_string(),
            modified_files: vec![],
        };
        let results = registry.execute_criteria(&criteria, &ctx);
        assert_eq!(results.len(), 2);
        assert_eq!(results[0].status, "passed");
        assert_eq!(results[1].status, "failed");
    }

    #[test]
    fn has_executor_registered() {
        let registry = ValidatorRegistry::new();
        assert!(registry.has_executor(&VerificationSpecType::FileExists));
        assert!(registry.has_executor(&VerificationSpecType::TestPass));
    }
}
